config/src/test/java/org/springframework/security/config/http/OAuth2LoginBeanDefinitionParserTests.java

```diff
@@ -0,0 +1,739 @@
+		this.spring.configLocations(this.xml("SingleClientRegistration-WithFormLogin")).autowire();
+
+		this.mvc.perform(get("/")).andExpect(status().is3xxRedirection())
+				.andExpect(redirectedUrl("http://localhost/login"));
+	}
+
+	@Test
+	public void customClientRegistrationRepository() throws Exception {
+		this.spring.configLocations(this.xml("WithCustomClientRegistrationRepository")).autowire();
+
+		assertThat(clientRegistrationRepository).isInstanceOf(MockClientRegistrationRepository.class);
+		ClientRegistration clientReg = clientRegistrationRepository.findByRegistrationId("mock");
+		assertThat(clientReg.getRegistrationId()).isEqualTo("mock");
+
+		Field field = oauth2LoginAuthenticationFilter.getClass().getDeclaredField("clientRegistrationRepository");
+		field.setAccessible(true);
+		Object fieldVal = field.get(oauth2LoginAuthenticationFilter);
+		assertThat(fieldVal).isInstanceOf(MockClientRegistrationRepository.class);
+		assertThat(fieldVal).isSameAs(clientRegistrationRepository);
+	}
+
+	@Test
+	public void customAuthorizedClientRepository() throws Exception {
+		this.spring.configLocations(this.xml("WithCustomAuthorizedClientRepository")).autowire();
+
+		assertThat(oauth2AuthorizedClientRepository).isInstanceOf(MockOAuth2AuthorizedClientRepository.class);
+
+		Field field = oauth2LoginAuthenticationFilter.getClass().getDeclaredField("authorizedClientRepository");
+		field.setAccessible(true);
+		Object fieldVal = field.get(oauth2LoginAuthenticationFilter);
+		assertThat(fieldVal).isInstanceOf(MockOAuth2AuthorizedClientRepository.class);
+		assertThat(fieldVal).isSameAs(oauth2AuthorizedClientRepository);
+	}
+
+	@Test
+	public void customAuthorizedClientService() throws Exception {
+		this.spring.configLocations(this.xml("WithCustomAuthorizedClientService")).autowire();
+
+		assertThat(oauth2AuthorizedClientService).isInstanceOf(MockOAuth2AuthorizedClientService.class);
+
+		Field authorizedClientRepositoryField = oauth2LoginAuthenticationFilter.getClass()
+				.getDeclaredField("authorizedClientRepository");
+		authorizedClientRepositoryField.setAccessible(true);
+		Object authorizedClientRepositoryFieldVal = authorizedClientRepositoryField
+				.get(oauth2LoginAuthenticationFilter);
+		assertThat(authorizedClientRepositoryFieldVal)
+				.isInstanceOf(AuthenticatedPrincipalOAuth2AuthorizedClientRepository.class);
+
+		Field authorizedClientServiceField = authorizedClientRepositoryFieldVal.getClass()
+				.getDeclaredField("authorizedClientService");
+		authorizedClientServiceField.setAccessible(true);
+		Object authorizedClientServiceFieldVal = authorizedClientServiceField.get(authorizedClientRepositoryFieldVal);
+		assertThat(authorizedClientServiceFieldVal).isInstanceOf(MockOAuth2AuthorizedClientService.class);
+		assertThat(authorizedClientServiceFieldVal).isSameAs(oauth2AuthorizedClientService);
+	}
+
+	@Test
+	public void customAuthorizationRequestRepository() throws Exception {
```

Comment:
Please update test to `verify()` that the mock `OAuth2AuthorizationRequestRepository` is called. We can also rename to `requestWhenCustomAuthorizationRequestRepositoryThenCalled`